### ISSUE TITLE
BLUS30767 pixel alignment or texture alpha problem

### ISSUE THREAD

**Issue by Nerboruto** [RPCS3/rpcs3#4133]
look at the attached picture

![rpcs3 2018-01-29 21-53-53](https://user-images.githubusercontent.com/10599012/35617117-99d36c0a-0677-11e8-87dc-0ba6f97d0cb4.png)

**Comment by Nerboruto** [comment]
even the shadows flash in that place...

**Comment by kd-11** [comment]
Just looks like texture border issues. Also, this report is lacking alot of information. What renderer is that? What settings? Does it happen on both opengl and vulkan? Does strict mode affect anything? etc

**Comment by Nerboruto** [comment]
opengl & vulkan on nvidia... 1280x720 scaling 100%.. monitor resolution 1360x768 + reshade custom scanlines. reshade & strict mode is irrelevant. the shadows become darker and blink. 
tested on 2 different nvidia PCs only at that point in the game. 
around the branches of the tree there is a black outline that appears
during the scrolling of the backdrop the problem tends to disappear and reappear and the shadows become more darker.

I would also have a suggestion for scaling resolution.
approximate resolutions to 16 pixels multiple.
es: 1280x720 @ 107% =1369x770 rounded at 1360x768.

**Comment by Nerboruto** [comment]
use gpu texture scaling is irrilevant. all options at default.

**Comment by Nerboruto** [comment]
it seems more a problem of how shadows are rendered over background texture. 
however I'm not sure.

**Comment by dio-gh** [comment]
Both the shadow flickering and the sprite misalignment issues are still present as of `RPCS3 v0.0.7-8852-e9ea226e Alpha`. I couldn't capture the shadow flicker, but the misalignment I did manage to; see the RRC file attached below. Captured on VK, replayed on GL, happens on both.

![image](https://user-images.githubusercontent.com/10439488/66614259-9b1f0a00-ebc8-11e9-89b2-0fd8071d11db.png)

RRC file: [BLUS30767_20191011014102_capture.zip](https://github.com/RPCS3/rpcs3/files/3715007/BLUS30767_20191011014102_capture.zip)

<details>
<summary>Guide to the area on the picture</summary>

The area can be accessed after about 5-10 minutes spent ingame. I also chose an Amazon character, as I wasn't sure how different the playthroughs are for each character archetype.

You have to do a quick tutorial that teaches you how to move around, fight, and interact with your environment. This takes about 5 minutes at worst.

Then you'll be forced to join a guild, but probably have yourself rejected for unkept armor, sent to repair it, and then will you be allowed to join. This takes about another minute.

After this, you're sent out to help a guild member, that's when you'll be able to access the area on the picture: it's called `Gate`. Going a bit left, and here we are.
</details>

**Comment by Nerboruto** [comment]
I think it's caused by shadows...
when the shadows flash textures edges turn black. it doesn't always happen but only while background move at some coordinates.

**Comment by digitaldude555** [comment]
@Nerboruto  update.

**Comment by Nerboruto** [comment]
the problem is still here...

**Comment by kd-11** [comment]
Need a new RSX capture, the one attached here is incompatible with current rpcs3.

**Comment by Nerboruto** [comment]
[BLUS30767_20200508224844_capture.zip](https://github.com/RPCS3/rpcs3/files/4601558/BLUS30767_20200508224844_capture.zip)

**Comment by Nerboruto** [comment]
[VID_20200508_230519.zip](https://github.com/RPCS3/rpcs3/files/4601640/VID_20200508_230519.zip)
shadows flikering

**Comment by 13xforever** [comment]
RenderDoc 1.7 [captures](https://1drv.ms/u/s!AruI8iDXabVJ4EoBrxJXX8fprQXr?e=a0IAdo) on current build 10398
I got four frames with various characters flickering, and most of them have texture alignment issues in the background sprites

**Comment by kd-11** [comment]
This is not caused by shadows or anything like that its a combination of factors. Unfortunately modern hardware does not behave the same way as a GPU from 20 years ago and some of the techniques this game is relying on are not present on modern cards anymore.
The borders are part of the artwork. You can see them in this pic (alpha channel isolated):
![image](https://user-images.githubusercontent.com/15904127/82139405-3697ed80-9830-11ea-82f8-ec8256b87da7.png)
But that's not all, if you zoom in closer, you get a better idea of where the line truly comes from:
![image](https://user-images.githubusercontent.com/15904127/82139441-6c3cd680-9830-11ea-8581-c81b03f8378b.png)
This is a compressed texture, so there is some 'bleeding' over the edges of alpha vs no alpha along with some banding due to low bit compression. So how does the game sample this texture? It uses the old-school coordinate clamp technique which no longer exists for over a decade (GL_CLAMP). From spec they just say "it clamps the coordinates to [0-1]" but I know that's not the whole story, it does some other fuckery too depending on the hardware. We can only enable this on old OGL v2 which has been obsolete since 2008. Rpcs3 has a workaround for this with GL_CLAMP_TO_EDGE, but its not the exact same behavior.
Whether the alpha border texels are decoded differently on a G70 class GPU is unknown, but not entirely out of the question either. But from what I can see, rpcs3 is doing the best it can with what we have to work with.

**Comment by Bertue** [comment]
Well, the transparent graphical errors of the paintings plus the cup in the menu in the RPCS3 game of Dragon's Crown continue to happen to me!

![screenshot_20220215_100727](https://user-images.githubusercontent.com/98273404/154029618-af821789-5f18-4a14-9f4c-bd1bea774075.png)

![screenshot_20220215_101244](https://user-images.githubusercontent.com/98273404/154030093-baed18fd-7de8-4dfd-b04f-422554a9c7c7.png)

**Comment by Nerboruto** [comment]
I just wanted to point out that the problem is present only when the background moves to certain coordinates not always

**Comment by dante3732** [comment]
Same, still not fixed.

**Comment by Peduls** [comment]
Seems like a duplicate of #9150

**Comment by Bertue** [comment]
If I think that the solution would be to expand the bilinear texture for the first time for 2D games like these, well, the kd-11 knows more about it, let's see what will be done sooner or later, but it will be less, I suppose, to be able to eliminate these absurd bugs of the Rpcs3 emulator and everyone is happy.

**Comment by Bertue** [comment]
I made a mistake, I recognize that I shouldn't have rushed, I'm sorry, when it has to happen, let it happen, please.

**Comment by Bertue** [comment]
Well, I think it is very difficult to fix these graphic bugs and more of old 2D games like these in the RPCS3 emulator and I have already seen why they fail due to the problem of OpenGL version 2.3 for much older computers that do not incorporate it and Vulkan I think that it does not reach him until they release some version of VK something that I have not seen at least that even if this kind of complicated fastening textures will work to solve the graphical errors of these two games because it is not even known what solution it has for what The texture clamping error would no longer exist in both Dragon's Crown and Odin Sphere because we have to thank those of Rpcs3, which is much better than the Vita3k emulator, which already works with those two games too, and they are much worse because of how green they are. is that emulator still!

**Comment by Bertue** [comment]
Well, I think it is very difficult to fix these graphic bugs and more of old 2D games like these in the RPCS3 emulator and I have already seen why they fail due to the problem of OpenGL version 2.3 for much older computers that do not incorporate it and Vulkan I think that it does not reach him until they release some version of VK something that I have not seen at least that even if this kind of complicated fastening textures will work to solve the graphical errors of these two games because it is not even known what solution it has for what The texture clamping error would no longer exist in both Dragon's Crown and Odin Sphere because we have to thank those of Rpcs3, which is much better than the Vita3k emulator, which already works with those two games too, and they are much worse because of how green they are. is that emulator still!

**Comment by Bertue** [comment]
Ogl bringup 4 #12284
I wish I could fix the few graphical glitches that both Dragon's Crown and Odin Sphere have, but I doubt it so soon yet.

**Comment by shinra-electric** [comment]
This should be fixed. 

<img width="1392" alt="Screenshot 2023-05-26 at 16 25 52" src="https://github.com/RPCS3/rpcs3/assets/50119606/f231b0a2-09a4-4317-8575-4a055462401d">

**Comment by 13xforever** [comment]
it's not fixed, and as kd explained before, the actual fix won't be easy or performant 
![image](https://github.com/RPCS3/rpcs3/assets/36445/1f24edce-7d54-49d4-b461-6e575b8f3d13)

**Comment by shinra-electric** [comment]
Ah, ok. I didn't realise it was a specific area. I can reproduce. 

<img width="1392" alt="Screenshot 2023-05-26 at 16 51 32" src="https://github.com/RPCS3/rpcs3/assets/50119606/d109d7bc-0847-4024-8060-7f2c38a1e1ad">

**Comment by kd-11** [comment]
Earlier this year I submitted some work that supports custom clamp algorithms (https://github.com/RPCS3/rpcs3/pull/14104)
I think we can now tackle this one in software as we have a pipeline to do it. I need a save file though and instructions how exactly to reproduce it (basically eli5)

**Comment by 13xforever** [comment]
You can reproduce all the issues from the beginning, it doesn't take much time to go through character creation and get to the tavern.

After you load save, leave the tavern and go all the way to the right, then exit through the Gate, and that's it.

On the first actual dungeon (Ancient Temple Ruins / Boss: Harpy Lv 1) if you go most of the way to the boss, there's a section with drowned ruins and a couple of lizard enemies where you could reproduce disappearing sprites (if you go too far right, you start the boss fight). But I think this was fixed already.

[dragon's crown save.zip](https://github.com/RPCS3/rpcs3/files/13761940/dragon.s.crown.save.zip)

**Comment by kd-11** [comment]
I finally got some time to look into this. It's a game bug for sure.
The game has a weird rendering setup and basically renders everything as tiles. The inputs are chunked and stitched together into texture atlases. This is fine, but introduces some 1px borders all over the place.
The borders are actually invisible, and are caused by a blending setup error. The developer set up the blend equation as S.src_alpha + D.(1 - src_alpha) for both color and alpha which is almost never what you want. However, the output is typically never sampled so nothing bad happens.
However, sometimes, to render special effects or menus, etc, the game can decide to duplicate a frame and resamples the previous frame. This is where everything falls apart, because now the alpha channel which is ignored normally is now relevant.

So, why doesn't this appear on PS3? Because they picked a special PS3 output format that erases the alpha channel. In rpcs3 this is implemented using a special mask when reading. However, the mask does not propagate properly when data is being copied around so we lose track of the "throw away the alpha" flag and we end up rendering the seams.

A fix is trivial, but unfortunately comes with a shader cache version bump. This also explains the weird issues observed in https://github.com/RPCS3/rpcs3/issues/9150

**Comment by kd-11** [comment]
Hmm, I'll try and come up with a less intrusive stop-gap solution for now.

**Comment by peilong-du** [comment]
Is this related to the weird boxes that show up in shadow-y areas in the game?

**Comment by Nightquaker** [comment]
Can confirm this issue is still present on RPCS3-0.0.38-18404. I'm playing on Linux Mint 22.2 using Vulkan, my GPU is AMD RX 5700 XT.

**Comment by persiopadre** [comment]
I can confirm that is present to this day, there's also some issue on darker areas with 'torch/fire' lighting

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/b24e7a25-0ec8-44cd-9b89-ab113ad7aeaa" />

<img width="1918" height="1080" alt="Image" src="https://github.com/user-attachments/assets/17fe1c24-bee3-48ab-b495-ff679ac89e4a" />

**Comment by keikof070623-gif** [comment]
Is this still an issue?

**Comment by persiopadre** [comment]
I can confirm that's still an issue up to 0.42-20039-56237b68 Alpha


<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/feda25a7-a038-46bd-b2e3-35cae992895d" />

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/ff040219-175b-4a9b-b7de-f9babc681390" />
